File: 01.Pattern-Sliding-Window/11.smallest_window_containing_substring/MinimumWindowSubstring.java
import java.util.*;

class MinimumWindowSubstring {
    public static String findSubstring(String str, String pattern) {
        int windowStart = 0, matched = 0, minLength = str.length() + 1, subStrStart = 0;
        Map<Character, Integer> charFrequenMap = new HashMap<>();
        for (char chr : pattern.toCharArray()) {
            charFrequenMap.put(chr, charFrequenMap.getOrDefault(chr, 0) + 1);
        }

        // try to extend the range [windowStart, windowEnd]
        for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
            char rightChar = str.charAt(windowEnd);
            if (charFrequenMap.containsKey(rightChar)) {
                charFrequenMap.put(rightChar, charFrequenMap.get(rightChar) - 1);
                if (charFrequenMap.get(rightChar) >= 0) { // count every matching of a character
                    matched++;
                }
            }

            // shrink the window if we can, finish as soon as we remove a matched character
            while (matched == pattern.length()) {
                if (minLength > windowEnd - windowStart + 1) {
                    minLength = windowEnd - windowStart + 1;
                    subStrStart = windowStart;
                }

                char leftChar = str.charAt(windowStart++);
                if (charFrequenMap.containsKey(leftChar)) {
                    // note that we could have redundant matching characters, thereforce we'll
                    // decrement the
                    // matched count only when a useful occurrence of a matched character is going
                    // out of the window
                    if (charFrequenMap.get(leftChar) == 0) {
                        matched--;
                    }

                    charFrequenMap.put(leftChar, charFrequenMap.get(leftChar) + 1);
                }
            }
        }

        return minLength > str.length() ? "" : str.substring(subStrStart, subStrStart + minLength);
    }

    public static void main(String[] args) {
        System.out.println(MinimumWindowSubstring.findSubstring("aabdec", "abc"));
        System.out.println(MinimumWindowSubstring.findSubstring("abdabca", "abc"));
        System.out.println(MinimumWindowSubstring.findSubstring("adcad", "abc"));
    }
}

File: 01.Pattern-Sliding-Window/11.smallest_window_containing_substring/minimum_window_substring.js
const findSubstring = function (str, pattern) {
  let windowStart = 0,
    matched = 0,
    substrStart = 0,
    minLength = str.length + 1,
    charFrequency = {};

  for (i = 0; i < pattern.length; i++) {
    const chr = pattern[i];
    if (!(chr in charFrequency)) {
      charFrequency[chr] = 0;
    }
    charFrequency[chr] += 1;
  }

  // try to extend the range [windowStart, windowEnd]
  for (windowEnd = 0; windowEnd < str.length; windowEnd++) {
    const rightChar = str[windowEnd];
    if (rightChar in charFrequency) {
      charFrequency[rightChar] -= 1;
      if (charFrequency[rightChar] >= 0) {
        // count avery matching of a character
        matched += 1;
      }
    }

    // Shrink the window if we can, finish as soon as we remove a matched character
    while (matched == pattern.length) {
      if (minLength > windowEnd - windowStart + 1) {
        minLength = windowEnd - windowStart + 1;
        substrStart = windowStart;
      }

      const leftChar = str[windowStart];
      windowStart++;
      if (leftChar in charFrequency) {
        // Note that we could have redundant matching characters, therefore we'll decrement the
        // matched count only when a useful occurrence of a matched character is going out of the window
        if (charFrequency[leftChar] === 0) {
          matched++;
        }
        charFrequency[leftChar] += 1;
      }
    }
  }

  if (minLength > str.length) {
    return "";
  }

  return str.substring(substrStart, substrStart + minLength);
};

console.log(findSubstring("aabdec", "abc"));
console.log(findSubstring("abdabca", "abc"));
console.log(findSubstring("adcad", "abc"));
